[10.0][ADD] cagnotte_sale: Add module to manage cagnotte on sale

rousseldenis opened the pull request:
Based on https://github.com/acsone/acsone-addons/pull/147

Commit:
add module cagnotte_base

Commit:
add module cagnotte_coupon

Commit:
add module cagnotte_coupon

Commit:
cagnotte : some fixes

Commit:
cagnotte : some fixes

Commit:
cagnotte_partner : set cagnotte on payment line

Commit:
add search on cagnotte and creation date

Commit:
add search on cagnotte and creation date

Commit:
add search on cagnotte and creation date

Commit:
cagnotte_base : ease the donation of coupon

Commit:
update security access

Commit:
update security access

Commit:
add active flag in unique constraint

Commit:
Prevent from cagnotte deletion if already used

Commit:
[MIG]port cagnotte modules to V10.0

Commit:
Flake8

Commit:
[FIX] fix wrong property name

Commit:
Code Review

Commit:
Fix wrong package name

Commit:
[CHG] improve unique constaint on account_cagnotte to apply only on active records

Commit:
[FIX] fix constraint on cagnotte unicity

Commit:
[FIX] fix wrong view mode on accounting entries

Commit:
[FIX] fix wrong anchor in xml view

Commit:
[FIX] fix view mode in cagnotte view

Commit:
[FIX] allow credit cagnotte

rousseldenis committed:
[10.0][FIX] cagnotte_base: Restore Monetary type for solde

rousseldenis committed:
[10.0][IMP] cagnotte_*: Improve tests, module description, ...

rousseldenis committed:
[10.0][FIX] .travis: Update to latest template

rousseldenis committed:
[10.0][ADD] cagnotte_sale: Add module to manage cagnotte on sale

Cedric-Pigeon commented on cagnotte_sale/models/sale_order.py:
```diff
@@ -0,0 +1,89 @@
+    @api.multi
+    def apply_cagnotte(self, cagnotte):
+        for sale in self:
+            if cagnotte.partner_id != sale.partner_id:
```
This won't work. The partner_id on the cagnotte is not mandatory.  When a customer buys a giftcard for someone else, no partner is set.

rousseldenis replied:
Ok, I had a doubt on that

Cedric-Pigeon commented on cagnotte_sale/models/sale_order.py:
```diff
@@ -0,0 +1,89 @@
+            if cagnotte.partner_id != sale.partner_id:
+                raise UserError(_('The cagnotte you try to use is not yours!'))
+            # We check that the sale order value is > 0
+            if cagnotte.solde_cagnotte > 0.0 and self.amount_total > 0.0:
```
would use `float_compare`

rousseldenis replied:
I'm not sure. I need to see in code how Odoo is treating Monetary fields (they are not pure float, so no decimal precision)

rousseldenis committed:
[10.0][IMP] cagnotte_sale: Change the solde computation, add wizard

As the balance of cagnotte should take into account not invoiced
sale orders, we need to identify them and deduct the applied
cagnotte amount to the cagnotte balance until the account move
lines are generated.

Add also a wizard to help paying with a cagnotte balance. This
wizard is accessible only for those in the good group.

rousseldenis committed:
[10.0][FIX] cagnotte_base: Don't use taxes on cagnotte product

rousseldenis committed:
[10.0][FIX] cagnotte_sale: Use write instead of inverse

As in interface, order lines are written through sale order object,
behaviour could be wrong.

rousseldenis committed:
[10.0][FIX] cagnotte_sale: Compute on sale amount total

Compute usable amount depending on sale order total tax included

Don't reapply cagnotte on sale order lines modifications to
avoid reentrant functions problems.

rousseldenis committed:
[10.0][FIX] cagnotte_sale: Reintroduce write() on sale order lines

rousseldenis committed:
[10.0][IMP] cagnotte_base: Create move line with invoice one

When the cagnotte is filled in on invoice line level,
the generated move is filled in with the cagnotte too.

rousseldenis committed:
[10.0][IMP] cagnotte_sale: Transmit cagnotte to invoice, fix write

Allows to transmit the cagnotte line to the corresponding invoice line.

Remove the write on sale order line: module limitation.

rousseldenis committed:
[10.0][IMP] cagnotte_sale: Improve name on sale order line

rousseldenis committed:
[10.0][IMP] cagnotte_partner: Add partner name if defined only

rousseldenis committed:
[10.0][FIX] cagnotte_sale: Compute solde on sale order changes

rousseldenis committed:
[10.0][IMP] cagnotte_base: Improve tests

rousseldenis committed:
[10.0][IMP] cagnotte_sale: Allows to use several cagnottes

rousseldenis committed:
[10.0][IMP] cagnotte_sale: Improve tree view
